parser-typechecker/src/Unison/Codebase/SqliteCodebase.hs

```diff
@@ -1121,43 +1118,20 @@ pushGitRootBranch srcConn branch repo = runExceptT @C.GitError do
       chId <- fromMaybe err <$> Q.loadCausalHashIdByCausalHash h2
       Q.setNamespaceRoot chId
 
-    -- This function makes sure that the result of git status is valid.
-    -- Valid lines are any of:
-    --
-    --   ?? .unison/v2/unison.sqlite3 (initial commit to an empty repo)
-    --   M .unison/v2/unison.sqlite3  (updating an existing repo)
-    --   D .unison/v2/unison.sqlite3-wal (cleaning up the WAL from before bugfix)
-    --   D .unison/v2/unison.sqlite3-shm (ditto)
-    --
-    -- Invalid lines are like:
-    --
-    --   ?? .unison/v2/unison.sqlite3-wal
-    --
-    -- Which will only happen if the write-ahead log hasn't been
-    -- fully folded into the unison.sqlite3 file.
-    --
-    -- Returns `Just (hasDeleteWal, hasDeleteShm)` on success,
-    -- `Nothing` otherwise. hasDeleteWal means there's the line:
-    --   D .unison/v2/unison.sqlite3-wal
-    -- and hasDeleteShm is `True` if there's the line:
-    --   D .unison/v2/unison.sqlite3-shm
-    --
-    parseStatus :: Text -> Maybe (Bool, Bool)
-    parseStatus status =
-      if all okLine statusLines then Just (hasDeleteWal, hasDeleteShm)
-      else Nothing
+    -- This function returns any lines of a git status which are invalid.
+    -- Valid lines are any modifications to .unison/v2/unison.sqlite3(-wal|-shm).
+    invalidStatusLines :: Text -> [Text]
```

Comment:
Since we're switching the remote DBs to all use WAL mode, it's okay if there are still WAL files floating around, we just have to be sure to manually checkpoint before pushing.

So I've adapted this function to only notice files other than the standard SQLite files.

Reply:
I’m still super unconvinced about switching around the transaction mode and eliminating this check.

Why are we switching the remote dbs to all use WAL mode again, more than they already were?

It seems to me that having multiple ucm users of a given git cache dir simultaneously has bigger concurrency issues than standard vs WAL transactions, or am I missing something?

Reply:
> Why are we switching the remote dbs to all use WAL mode again, more than they already were?

Would you mind enumerating your objections to using WAL mode so I can better address them please?

Here are the pros [straight from the sqlite docs](https://www.sqlite.org/wal.html)
1. WAL is significantly faster in most scenarios.
2. WAL provides more concurrency as readers do not block writers and a writer does not block readers. Reading and writing can proceed concurrently.
3. Disk I/O operations tends to be more sequential using WAL.
4. WAL uses many fewer fsync() operations and is thus less vulnerable to problems on systems where the fsync() system call is broken.

The primary reason I make the switch here is point two; in combination with the fact that we weren't properly closing database connections in many cases. Yes, we should probably just fix the closing logic, but I think the improved concurrency is still a boon here.

An additional benefit is just the conceptual simplicity of using the same mode for all of our databases.

The only real Con that I think applies to us from that page:
* There is the extra operation of checkpointing which, though automatic by default, is still something that application developers need to be mindful of.

Which is true, but we can simply checkpoint before pushing, so I think we're fine on that point.

If you remain super unconvinced, can you explain why please, and maybe also try to convince me and @mitchellwrosen  as to why DELETE mode is preferred?

Reply:
We already do use WAL mode, to get the benefits you listed.  Any time we open a local or remote codebase, we use WAL.

As of #2064, ucm switches the _git cache_ database's mode to DELETE before pushing, a time when there are no anticipated other connections, to work around the fact that in some unidentified code paths, the database apparently wasn't being closed and cleaned up correctly.  We want sqlite to agree that the db file is in a good state, with its journal fully committed before adding it to git. The last-second switch to DELETE mode does this and cleans up the external journal files.

If there are multiple processes accessing the a git cache database during a push, that's something we need to understand and resolve independently of which journal mode is selected.

Additionally, if we want to reevaluate and change all of that, I think it should be a separate PR from "don't hard-reset git repos if there are no remote changes", should be called out in the PR description as well, and should get significant testing (documented in the PR, if manual testing).

Reply:
> We already do use WAL mode to get the benefits you listed. Any time we open a local or remote codebase, we use WAL.

I understand that, I think it's worth having those benefits for git repo codebases too 😄 

> a time when there are no anticipated other connections

While I agree it's less likely that repos will be accessed concurrently, it's definitely still possible. I'd prefer that ucm is safe to use concurrency in all cases, not just _mostly_ safe; otherwise we'll get rare and confusing bug reports of crashes in the future without necessarily knowing where they're coming from.

>If there actually are multiple processes accessing the a git cache database during a push, that's something we need to understand and resolve independently of which journal mode is selected.

Is there more to understand here other than that more than one ucm may be pulling/pushing at a time? I admit that more than one push at a time is strange and troublesome, but multiple updates to the git repo codebase should be allowed in general I think. In general, treating the database mode as a connection-local concept by switching it according to the task at hand seems maladaptive to concurrency to me.

At any rate, I've tried leaving it in `DELETE` mode with the close-connections and the tests still continue to fail 😢 
so if this is a blocker then we'll have to decide between:

1. taking more time to understand why the tests are failing (probably a good idea regardless)
2. leaving off this behaviour and re-migrating git-repos on every sync/pull
3. finding some other way to achieve this behaviour.

Reply:
> > We already do use WAL mode to get the benefits you listed. Any time we open a local or remote codebase, we use WAL.
> 
> I understand that, I think it's worth having those benefits for git repo codebases too 😄

I'm not sure if you mean the git repo codebases as they sit on the git servers, but when we clone them and access them locally we use WAL mode.

> > a time when there are no anticipated other connections
> 
> While I agree it's less likely that repos will be accessed concurrently, it's definitely still possible. I'd prefer that ucm is safe to use concurrency in all cases, not just _mostly_ safe; otherwise we'll get rare and confusing bug reports of crashes in the future without necessarily knowing where they're coming from.

Agreed, but I think relying on WAL to make a single-threaded test pass gets us further from safe concurrent use in all cases, rather than closer to it.

> > If there actually are multiple processes accessing the a git cache database during a push, that's something we need to understand and resolve independently of which journal mode is selected.
> 
> Is there more to understand here other than that more than one ucm may be pulling/pushing at a time? I admit that more than one push at a time is strange and troublesome, but multiple updates to the git repo codebase should be allowed in general I think. In general, treating the database mode as a connection-local concept by switching it according to the task at hand seems maladaptive to concurrency to me.

I think so.  

1. I don't know off the top of my head what multiple `git` and `ucm` are doing to the working dir or to the database when pushing/pulling at the same time.
2. I don't know what state the database is in if we try to stage and push it in the middle of concurrent access.
3. I don't think we should be pushing partially written journals to Github.
4. I don't think we should be not pushing partially written journals to Github.
5. I understand there a sync command we can send to sqlite that writes the journals into the database, but there's going to be a race condition between doing that and staging and pushing with git.
6. Maybe others?

> At any rate, I've tried leaving it in `DELETE` mode with the close-connections and the tests still continue to fail 😢 so if this is a blocker then we'll have to decide between:
> 
> 1. taking more time to understand why the tests are failing (probably a good idea regardless)
> 2. leaving off this behaviour and re-migrating git-repos on every sync/pull
> 3. finding some other way to achieve this behaviour.

These all sound good.  We can certainly start with (2) while pursuing (1) and (3); (1) sounds best to me.